WIP: Reproducibility note

themightyoarfish opened the pull request:
This adds a Note on making experiments reproducible. 

It also adds Instructions for building the Documentation to `README.md`. Please ping if I missed any requirements.

I'm not sure what to do about the submodule changes. Please advise.

Commit:
Add instructions for building the documentation locally

Commit:
Add note on reproducibility

t-vi commented:
I think there are known reproducibility issues with distributed (search the forums).
Also things like index_add (using atomic add) are not deterministic, this includes CTC loss - although the latter might possibly be fixable by fixing the precision of the atomic add operands.

themightyoarfish commented:
Ok then the note should be amended, but would you agree that there should be some documentation on it? At least on a single machine this would do the trick, and finding this information was really hard.

SsnL commented:
@themightyoarfish Please revert the submodule change. Moreover, it might be good to mention that reproducibility is not guaranteed across versions and/or architectures. E.g., a RNG on CUDA definitely behaves different than the RNG on CPU even though they use same seed; and when we optimize a sampling kernel in a release, the generated numbers might be different (although distribution is the same).

Commit:
Improve note on randomness

- Add caveat that reproducibility is not guaranteed across platforms or
releases

ezyang commented on docs/source/notes/randomness.rst:
```diff
@@ -0,0 +1,41 @@
+
+Numpy
+.....
+The numpy rng can be seeded with::
```
I wouldn't put the bit about Numpy here; not everyone is going to use Numpy. I'd put it after PyTorch, and preface it with, "if you, or any of the libraries you are using, use numpy..."

themightyoarfish replied:
Good suggestion; I just searched for `numpy` in the repo and deduced from that that numpy is used somewhere within PyTorch.

ezyang commented on docs/source/notes/randomness.rst:
```diff
@@ -0,0 +1,41 @@
+
+While this will seed all CUDA RNGs as well, you can do so explicitly with
+:meth:`torch.cuda.manual_seed` for the current device, or
+:meth:`torch.cuda.manual_seed_all` to seed the RNG for all GPUs
```
It's worth noting why you might *not* want to `manual_seed_all`, which is that the time it takes scales with the number of GPUs you seed, so if you have a lot of GPUs but are only using one of them, `manual_seed_all` will needlessly bloat your load time,

soumith replied:
`torch.manual_seed` will set all devices lazily, including all CUDA GPUs.
It USED to be that the behavior of `torch.manual_seed` only seeded CPU.
Hence, this section can simply be said: `To seed all devices with a fixed RNG, set `torch.manual_seed(0)`.

In fact, @ezyang wrote that patch :)

Commit:
Update note on randomness

soumith commented on README.md:
````diff
@@ -224,6 +225,17 @@ Please note that PyTorch uses shared memory to share data between processes, so
+To build documentation in various formats, you will need Sphinx.
+
+```
+conda install sphinx sphinx_rtd_theme
````
this should be:

```
cd docs/
pip install -r requirements.txt
```

themightyoarfish replied:
Huh, totally did not see that that exists. Fixing.

Commit:
Fix doc build instructions

themightyoarfish commented:
Seems like we have a typo in the note
`Furthermore, results need to be
reproducible between CPU and GPU executions`, there's a `not` missing here.

themightyoarfish commented:
@ezyang Fixed the problem in #11571
